File: game/map-info.go
package game

import (
	"fmt"
	"math"
)

type MapVNodeNeighborIdx struct {
	TopLeft      uint `json:"topLeft"`
	TopMiddle    uint `json:"topMiddle"`
	TopRight     uint `json:"topRight"`
	MiddleRight  uint `json:"middleRight"`
	BottomRight  uint `json:"bottomRight"`
	BottomMiddle uint `json:"bottomMiddle"`
	BottomLeft   uint `json:"bottomLeft"`
	MiddleLeft   uint `json:"middleLeft"`
}

type MapNode struct {
	G           uint                `json:"g"`
	H           uint                `json:"h"`
	VPosition   UPoint              `json:"position"`
	RPosition   Point               `json:"mapPosition"`
	ParentIdx   *uint               `json:"partnerIdx"`
	NeighborIdx MapVNodeNeighborIdx `json:"neighborIdx"`
	Walkable    bool                `json:"walkable"`
}

func (node *MapNode) F() uint {
	return node.G + node.H
}

func (node *MapNode) SetParams(start UPoint, end UPoint) {
	node.G = MapInfoDistanceBetweenPoints(start, node.VPosition)
	node.H = MapInfoDistanceBetweenPoints(node.VPosition, end)
}

type MapRVTransform struct {
	X MathFnLinear2D `json:"x"`
	Y MathFnLinear2D `json:"y"`
}

type MapInfo struct {
	Nodes       []MapNode      `json:"nodes"`       // Virtual node of size 1x1
	VSize       USize          `json:"size"`        // Virtual map size, can be filled with whole number of VNodes
	RSize       Size           `json:"mapSize"`     // Real map size, as it appears in the game
	RNodeSize   Size           `json:"mapNodeSize"` // Size of the VNode on the real map
	RVTransform MapRVTransform `json:"rvTransform"` // Transform function to get from V space to R space and vice versa
}

func NewMapInfo(pixels []uint8, vSize USize, rSize Size) MapInfo {
	var nodes []MapNode
	rvTransform := MapRVTransform{
		X: NewMathFnLinear2D([]Point{
			{
				X: 0.0,
				Y: 0.0,
			},
			{
				X: rSize.Width,
				Y: float64(vSize.Width),
			},
		}),
		Y: NewMathFnLinear2D([]Point{
			{
				X: 0.0,
				Y: 0.0,
			},
			{
				X: rSize.Height,
				Y: float64(vSize.Height),
			},
		}),
	}
	var x uint = 0
	var y uint = 0
	rNodeSize := Size{
		Width:  rvTransform.X.Inverse(1.0),
		Height: rvTransform.Y.Inverse(1.0),
	}
	for i := 0; i < len(pixels); i++ {
		node := MapNode{
			G:         0,
			H:         0,
			VPosition: NewUPoint(x, y),
			RPosition: Point{
				X: rvTransform.X.Inverse(float64(x)) + rNodeSize.Width/2.0,
				Y: rvTransform.Y.Inverse(float64(y)) + rNodeSize.Height/2.0,
			},
			ParentIdx: nil,
			NeighborIdx: MapVNodeNeighborIdx{
				1000000000,
				1000000000,
				1000000000,
				1000000000,
				1000000000,
				1000000000,
				1000000000,
				1000000000,
			},
			Walkable: false,
		}
		if pixels[i] < 200 {
			node.Walkable = true
		}
		if x > 0 && y > 0 {
			node.NeighborIdx.TopLeft = (x - 1) + (vSize.Width * (y - 1))
		}
		if y > 0 {
			node.NeighborIdx.TopMiddle = x + (vSize.Width * (y - 1))
		}
		if x < vSize.Width-1 && y > 0 {
			node.NeighborIdx.TopRight = (x + 1) + (vSize.Width * (y - 1))
		}
		if x < vSize.Width-1 {
			node.NeighborIdx.MiddleRight = (x + 1) + (vSize.Width * y)
		}
		if x < vSize.Width-1 && y < vSize.Height-1 {
			node.NeighborIdx.BottomRight = (x + 1) + (vSize.Width * (y + 1))
		}
		if y < vSize.Height-1 {
			node.NeighborIdx.BottomMiddle = x + (vSize.Width * (y + 1))
		}
		if x > 0 && y < vSize.Height-1 {
			node.NeighborIdx.BottomLeft = (x - 1) + (vSize.Width * (y + 1))
		}
		if x > 0 {
			node.NeighborIdx.MiddleLeft = (x - 1) + (vSize.Width * y)
		}
		nodes = append(nodes, node)
		x++
		if x == vSize.Width {
			x = 0
			y++
			if y == vSize.Height {
				break
			}
		}
	}
	return MapInfo{
		nodes,
		vSize,
		rSize,
		rNodeSize,
		rvTransform,
	}
}

func (info *MapInfo) GetValidNode(rPosition Point) (*MapNode, bool) {
	x := uint(info.RVTransform.X.Calc(rPosition.X))
	y := uint(info.RVTransform.Y.Calc(rPosition.Y))
	nodeIdx := x + info.VSize.Width*y
	if nodeIdx < uint(len(info.Nodes)) {
		if info.Nodes[nodeIdx].Walkable {
			return &info.Nodes[nodeIdx], true
		} else {
			doLoop := true
			offset := []int{-1, 1}
			for doLoop {
				foundValidNode := false
				for xOffs := offset[0]; xOffs < offset[1]; xOffs++ {
					for yOffs := offset[0]; yOffs < offset[1]; yOffs++ {
						x = info.Nodes[nodeIdx].VPosition.X + uint(xOffs)
						y = info.Nodes[nodeIdx].VPosition.Y + uint(yOffs)
						nodeIdx = x + info.VSize.Width*y
						if nodeIdx < uint(len(info.Nodes)) {
							foundValidNode = true
							if info.Nodes[nodeIdx].Walkable == true {
								return &info.Nodes[nodeIdx], false
							}
						}
					}
				}
				if foundValidNode == false {
					doLoop = false
				}
				offset[0] -= 1
				offset[1] += 1
			}
		}
	}
	return nil, false
}

func (info *MapInfo) GetNodeAtVPosition(vPosition UPoint) *MapNode {
	for i := 0; i < len(info.Nodes); i++ {
		if info.Nodes[i].VPosition.X == vPosition.X && info.Nodes[i].VPosition.Y == vPosition.Y {
			return &info.Nodes[i]
		}
	}
	return nil
}

func MapInfoDistanceBetweenPoints(start UPoint, end UPoint) uint {
	x := uint(math.Abs(float64(end.X - start.X)))
	y := uint(math.Abs(float64(end.Y - start.Y)))
	if x > y {
		return 14*y + 10*(x-y)
	}
	return 14*x + 10*(y-x)
}

func (api *Api) MapInfoSet(pixels []uint8, mapSlug string) MapInfo {
	mapData := StateData.FindMap(mapSlug)
	if mapData == nil {
		panic(fmt.Sprintf("Map with slug %s does not exist", mapSlug))
	}
	StateData.MapInfo = NewMapInfo(pixels, USize{
		Width:  uint(mapData.Nogo.Width),
		Height: uint(mapData.Nogo.Height),
	}, Size{
		Width:  mapData.Width,
		Height: mapData.Height,
	})
	return StateData.MapInfo
}

File: src-tauri/src/game/map_info.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::game::map_node::{MapNode, MapNodeNeighborIdx};
use crate::game::point::{Point, UPoint};
use crate::game::size::{Size, USize};
use crate::GameState;

use super::math::MathFnLinear2D;

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[ts(export)]
pub struct MapInfoRVTransform {
    pub x: MathFnLinear2D,
    pub y: MathFnLinear2D,
}

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[ts(export)]
pub struct MapInfo {
    pub nodes: Vec<MapNode>,
    pub v_size: USize,
    // Virtual map size
    pub r_size: Size,
    // Real map size
    pub r_node_size: Size,
    // Real node size on the map, in v_size node is always 1x1
    pub rv_transform: MapInfoRVTransform,
}

impl MapInfo {
    pub fn distance_between_points(start: &UPoint, end: &UPoint) -> usize {
        let x = end.x.abs_diff(start.x);
        let y = end.y.abs_diff(start.y);
        if x > y {
            return 14 * y + 10 * (x - y);
        }
        return 14 * x + 10 * (y - x);
    }

    pub fn new(
        pixels: Vec<u8>,
        v_size: USize,
        r_size: Size,
    ) -> MapInfo {
        let mut nodes: Vec<MapNode> = vec![];
        let mut i = 0;
        let mut x: usize = 0;
        let mut y: usize = 0;
        let rv_transform = MapInfoRVTransform {
            x: MathFnLinear2D::new(vec![Point::new(0.0, 0.0), Point::new(r_size.width, v_size.width as f32)]),
            y: MathFnLinear2D::new(vec![Point::new(0.0, 0.0), Point::new(r_size.height, v_size.height as f32)]),
        };
        let r_node_size = Size::new(rv_transform.x.inverse(1.0), rv_transform.y.inverse(1.0));
        while i < pixels.len() {
            let mut node = MapNode {
                v_position: UPoint::new(x, y),
                r_position: Point::new(rv_transform.x.inverse(x as f32) + r_node_size.width / 2.0, rv_transform.y.inverse(y as f32) + r_node_size.height / 2.0),
                g: 0,
                h: 0,
                parent_idx: None,
                walkable: false,
                neighbor_idx: MapNodeNeighborIdx::new(
                    1000000000,
                    1000000000,
                    1000000000,
                    1000000000,
                    1000000000,
                    1000000000,
                    1000000000,
                    1000000000,
                ),
            };
            if pixels[i] < 200 {
                node.walkable = true;
            }
            if x > 0 && y > 0 {
                node.neighbor_idx.top_left = ((x - 1) + (v_size.width * (y - 1))) as usize;
            }
            if y > 0 {
                node.neighbor_idx.top_mid = ((x - 0) + (v_size.width * (y - 1))) as usize;
            }
            if x < v_size.width - 1 && y > 0 {
                node.neighbor_idx.top_right = ((x + 1) + (v_size.width * (y - 1))) as usize;
            }
            if x < v_size.width - 1 {
                node.neighbor_idx.mid_right = ((x + 1) + (v_size.width * (y - 0))) as usize;
            }
            if x < v_size.width - 1 && y < v_size.height - 1 {
                node.neighbor_idx.bottom_right = ((x + 1) + (v_size.width * (y + 1))) as usize;
            }
            if y < v_size.height - 1 {
                node.neighbor_idx.bottom_mid = ((x - 0) + (v_size.width * (y + 1))) as usize;
            }
            if x > 0 && y < v_size.height - 1 {
                node.neighbor_idx.bottom_left = ((x - 1) + (v_size.width * (y + 1))) as usize;
            }
            if x > 0 {
                node.neighbor_idx.mid_left = ((x - 1) + (v_size.width * (y - 0))) as usize;
            }
            nodes.push(node);
            i += 1;
            x += 1;
            if x == v_size.width {
                x = 0;
                y += 1;
                if y == v_size.height {
                    break;
                }
            }
        }
        MapInfo {
            nodes,
            v_size,
            r_size,
            rv_transform,
            r_node_size,
        }
    }

    pub fn get_valid_node(
        &self,
        map_position: &Point,
    ) -> (
        Option<&MapNode>, // Node
        bool,                     // Is node found at target position or close to target
    ) {
        let x = self.rv_transform.x.calc(map_position.x) as usize;
        let y = self.rv_transform.y.calc(map_position.y) as usize;
        let chunk_idx = x + self.v_size.width * y;
        if chunk_idx < self.nodes.len() {
            if self.nodes[chunk_idx].walkable {
                return (Some(&self.nodes[chunk_idx]), true);
            } else {
                let mut offset: (i32, i32) = (-1, 1);
                let mut do_loop = true;
                while do_loop {
                    let mut found_valid_node = false;
                    for x_offs in offset.0..offset.1 {
                        for z_offs in offset.0..offset.1 {
                            let x = self.nodes[chunk_idx].v_position.x as i32 + x_offs;
                            let z = self.nodes[chunk_idx].v_position.y as i32 + z_offs;
                            let node_idx = (x as usize) + (self.v_size.width as usize * z as usize);
                            if node_idx < self.nodes.len() {
                                found_valid_node = true;
                                if self.nodes[node_idx].walkable == true {
                                    return (Some(&self.nodes[node_idx]), false);
                                }
                            }
                        }
                    }
                    if found_valid_node == false {
                        do_loop = false;
                    }
                    offset.0 -= 1;
                    offset.1 += 1;
                }
            }
        }
        (None, false)
    }

    pub fn get_node_at_position(&self, position: UPoint) -> Option<MapNode> {
        for i in 0..self.nodes.len() {
            if position.x == self.nodes[i].v_position.x && position.y == self.nodes[i].v_position.y {
                return Some(self.nodes[i].clone());
            }
        }
        None
    }
}

#[tauri::command]
pub fn map_info_set(state: tauri::State<GameState>, pixels: Vec<u8>, map_slug: &str) -> MapInfo {
    let mut state_guard = state.0.lock().unwrap();
    let map_data_opt = state_guard.find_map(map_slug);
    match map_data_opt {
        Some(map_data) => {
            state_guard.map_info = MapInfo::new(
                pixels,
                USize::new(
                    map_data.nogo.width as usize,
                    map_data.nogo.height as usize,
                ),
                Size::new(
                    map_data.width as f32,
                    map_data.height as f32,
                ),
            );
            return state_guard.map_info.clone();
        }
        None => {
            panic!("Map info: Map '{}' does not exist.", map_slug);
        }
    }
}
